src/pages/VeSDL/OnChainVote.tsx

```diff
@@ -0,0 +1,286 @@
+import VoteHistory from "./Votes"
+import { enqueuePromiseToast } from "../../components/Toastify"
+import { useActiveWeb3React } from "../../hooks"
+import { useTranslation } from "react-i18next"
+
+interface OnChainVoteProps {
+  veSdlBalance: BigNumber
+  gauges: Partial<{
+    [lpTokenAddress: string]: Gauge
+  }>
+  gaugeControllerContract: GaugeController
+}
+export type GaugeName = {
+  [gaugeAddress: string]: {
+    address: string
+    gaugeName: string
+  }
+}
+
+export type VotesType = {
+  [gaugeAddress: string]: {
+    gaugeName: string
+    weight: BigNumber
+    voteDate: number
+  }
+}
+
+export default function OnChainVote({
+  veSdlBalance,
+  gauges,
+  gaugeControllerContract,
+}: OnChainVoteProps) {
+  const theme = useTheme()
+  const { account } = useActiveWeb3React()
+  const { t } = useTranslation()
+  const [selectedGauge, setSelectedGauge] = useState<{
+    gaugeName: string
+    address: string
+  } | null>(null)
+  const [alertMessage, setAlertMessage] = useState<string>()
+  const [voteWeightToSubmit, setVoteWeightToSubmit] = useState<string>("")
+  const [voteUsed, setVoteUsed] = useState<number | undefined>()
+  const [votes, setVotes] = useState<VotesType>({})
+  const [voteWeightInputError, setVoteWeightInputError] = useState<{
+    hasError: boolean
+    errorText: string
+  }>({ hasError: false, errorText: " " })
+
+  const gaugeNames = useMemo(() => {
+    const gaugeNameObj = {} as GaugeName
+    Object.keys(gauges)
+      .filter((lpAddress) => !gauges[lpAddress]?.isKilled)
+      .forEach((lpAddress) => {
+        const gaugeName = gauges[lpAddress]?.gaugeName
+        const gaugeAddress = gauges[lpAddress]?.address.toLowerCase()
+        if (gaugeName && gaugeAddress)
+          gaugeNameObj[gaugeAddress] = {
+            address: gaugeAddress,
+            gaugeName: gaugeName,
+          }
+      })
+    return gaugeNameObj
+  }, [gauges])
+
+  const getVoteUsed = useCallback(async () => {
+    if (account) {
+      const voteUsed = await gaugeControllerContract[
+        "vote_user_power(address)"
+      ](account)
+      setVoteUsed(voteUsed.toNumber())
+    }
+  }, [account, gaugeControllerContract])
+
+  const getFilteredVotes = useCallback(async () => {
+    const voteFilter = gaugeControllerContract.filters.VoteForGauge(
+      null,
+      null,
+      null,
+      null,
+    )
+    const filteredRes = await gaugeControllerContract.queryFilter(voteFilter)
+
+    const filteredVotes = filteredRes.reduce(
+      (acc, { args: { user, time, weight, gauge_addr } }) => {
+        if (user === account) {
+          acc[gauge_addr.toLowerCase()] = {
+            gaugeName: gaugeNames[gauge_addr.toLowerCase()].gaugeName,
+            voteDate: time.toNumber(),
+            weight,
+          }
+        }
+        return acc
+      },
+      {} as VotesType,
+    )
+    setVotes(filteredVotes)
+  }, [account, gaugeControllerContract, gaugeNames])
+
+  useEffect(() => {
```

Comment:
instead of leveraging useCallback + useEffect can you use react-query?

Reply:
updated